rpc/resource/resource.go

```diff
@@ -0,0 +1,114 @@
+package resource //import "github.com/infobloxopen/atlas-app-toolkit/rpc/resource"
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/protobuf/proto"
+
+	"github.com/infobloxopen/atlas-app-toolkit/rpc/resource/resourcepb"
+)
+
+const defaultResource = "<default>"
+
+var (
+	mu       sync.RWMutex
+	registry = make(map[string]Codec)
+)
+
+// Identifier defines an interface to represent RPC Identifier.
+type Identifier interface {
+	// ApplicationName is an application identifier that will be used among
+	// other infrastructure services to identify the application
+	ApplicationName() string
+
+	// ResourceType is an application specific type name of a resource
+	ResourceType() string
+
+	// ResourceID is an application specific resource identity of a resource
+	ResourceID() string
+
+	// External is a flag indicates whether the identifier represents
+	// an external resource or holds an identity of a resource that is belongs
+	// to an application itself.
+	External() bool
+
+	// Valid is a flag indicates whether the identifier is a valid or not.
+	// The criteria is up to an implementation.
+	Valid() bool
+}
+
+// Codec defines the interface package uses to encode and decode RPC Identifier
+// to an Identifier interface implementation.
+// Note that implementation must be thread safe.
+type Codec interface {
+	// Encode returns RPC representation of Identifier interface
+	Encode(Identifier) (*resourcepb.Identifier, error)
+	// Decode returns implementation of Identifier interface based on RPC representation
+	Decode(*resourcepb.Identifier) (Identifier, error)
+}
+
+// RegisterCodec registers codec for a given pb.
+// If pb is nil the codec is registered as default.
+// If codec is nil or registered twice for the same resource
+// the panic is raised.
```

Comment:
why panic instead of return `error`?

Reply:
Registration happens at start of an application and should be defined somewhere in the code e.g in `init` functions. If an developer is registering the same codec twice for the same resource it seems it a bug.

It is done to indicate that a codec is overrided. 

It is unlikely that someone would want to encode/decode identifiers at runtime using one codec and then after some period change the codec to another one.

Reply:
one idea if it should happen at application startup only is that it is only accessible via a constructor instead of as a function. That way it's impossible to abuse the functionality.

Also, I see you're using a global singleton pattern here, which is somewhat concerning... I guess Go does it liberally as well, but I'm struggling to not see it as an antipattern.

Reply:
Ok, what do you propose?

Reply:
don't know yet... just throwing out unqualified opinions at this point :)